File: backend/app/schemas/inbox.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field, model_validator


class InboxSubmitRequest(BaseModel):
    url: str | None = None
    title: str | None = None
    content: str | None = None
    source_id: UUID | None = None

    @model_validator(mode="after")
    def require_at_least_one_field(self) -> "InboxSubmitRequest":
        if not any(
            value and str(value).strip()
            for value in (self.url, self.title, self.content)
        ):
            raise ValueError("At least one of url, title, or content is required.")
        return self


class InboxRawItem(BaseModel):
    id: UUID
    workspace_id: UUID | None
    title: str
    url: str | None
    content: str | None
    source_id: UUID
    content_hash: str
    published_at: datetime | None
    captured_at: datetime


class InboxSubmitData(BaseModel):
    raw_item: InboxRawItem
    duplicate: bool = False


class InboxListItem(BaseModel):
    raw_item: InboxRawItem
    analysis_status: Literal["pending", "complete"]
    has_intel_file: bool = False


class InboxListData(BaseModel):
    items: list[InboxListItem]
    total: int
    page: int
    page_size: int = Field(default=20)

File: apps/web/src/types/inbox.ts
export type ApiError = {
  code: string;
  message: string;
};

export type ApiResponse<T> = {
  success: boolean;
  data: T | null;
  error: ApiError | null;
};

export type InboxRawItem = {
  id: string;
  title: string;
  url: string | null;
  content: string | null;
  source_id: string;
  content_hash: string;
  published_at: string | null;
  captured_at: string;
};

export type InboxSubmitData = {
  raw_item: InboxRawItem;
  duplicate: boolean;
};

export type InboxListItem = {
  raw_item: InboxRawItem;
  analysis_status: "pending" | "complete";
  has_intel_file: boolean;
};

export type InboxListData = {
  items: InboxListItem[];
  total: number;
  page: number;
  page_size: number;
};

export type InboxSubmitRequest = {
  url?: string;
  title?: string;
  content?: string;
  source_id?: string | null;
};

export type SignalAnalysis = {
  id: string;
  raw_item_id: string;
  summary: string;
  signal_type: string;
  entities: unknown[];
  keywords: string[];
  claims: unknown[];
  suggested_tracking_queries: string[];
  novelty_score: number | null;
  relevance_score: number | null;
  credibility_hint: number | null;
  risk_hint: number | null;
  opportunity_types: string[];
  rationale: string | null;
  model: string | null;
  prompt_version: string | null;
  created_at: string;
};

export type AnalyzeData = {
  analysis: SignalAnalysis;
};
